exchanges/kucoin/kucoin_test.go

```diff
@@ -0,0 +1,2419 @@
+	_, err := ku.GetServerTime(context.Background(), asset.Spot)
+	if err != nil {
+		t.Error(err)
+	}
+	_, err = ku.GetServerTime(context.Background(), asset.Futures)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetRecentTrades(t *testing.T) {
+	t.Parallel()
+	enabledPairs, err := ku.GetEnabledPairs(asset.Futures)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = ku.GetRecentTrades(context.Background(), enabledPairs[0], asset.Futures)
+	if err != nil {
+		t.Error(err)
+	}
+	enabledPairs, err = ku.GetEnabledPairs(asset.Spot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = ku.GetRecentTrades(context.Background(), enabledPairs[0], asset.Spot)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetOrderHistory(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	var getOrdersRequest = order.GetOrdersRequest{
+		Type:      order.Limit,
+		Pairs:     []currency.Pair{currency.NewPair(currency.LTC, currency.BTC), currency.NewPair(currency.MHC, currency.ETH), currency.NewPair(currency.MHC, currency.BTC), currency.NewPair(currency.OXEN, currency.BTC)},
+		AssetType: asset.Futures,
+		Side:      order.Sell,
+	}
+	_, err := ku.GetOrderHistory(context.Background(), &getOrdersRequest)
+	if err != nil {
+		t.Error(err)
+	}
+	getOrdersRequest.Pairs = []currency.Pair{}
+	_, err = ku.GetOrderHistory(context.Background(), &getOrdersRequest)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetActiveOrders(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	enabledPairs, err := ku.GetEnabledPairs(asset.Spot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var getOrdersRequest = order.GetOrdersRequest{
+		Type:      order.Limit,
+		Pairs:     enabledPairs,
+		AssetType: asset.Futures,
+		Side:      order.Buy,
+	}
+	if _, err = ku.GetActiveOrders(context.Background(), &getOrdersRequest); err != nil {
+		t.Error("Kucoin GetActiveOrders() error", err)
+	}
+	getOrdersRequest.Pairs = []currency.Pair{}
+	if _, err = ku.GetActiveOrders(context.Background(), &getOrdersRequest); err != nil {
+		t.Error("Kucoin GetActiveOrders() error", err)
+	}
+}
+
+func TestGetFeeByType(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	if _, err := ku.GetFeeByType(context.Background(), &exchange.FeeBuilder{
+		Amount:              1,
+		FeeType:             exchange.CryptocurrencyTradeFee,
+		Pair:                currency.NewPairWithDelimiter(currency.BTC.String(), currency.USDT.String(), currency.DashDelimiter),
+		PurchasePrice:       1,
+		FiatCurrency:        currency.USD,
+		BankTransactionType: exchange.WireTransfer,
+	}); err != nil {
+		t.Errorf("%s GetFeeByType() error %v", ku.Name, err)
+	}
+}
+
+func TestValidateCredentials(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	if ku.ValidateCredentials(context.Background(), asset.Spot) != ku.ValidateCredentials(context.Background(), asset.Futures) {
```

Comment:
I'm not sure how this test is supposed to work sorry. KuCoin does not share credentials between SPOT and Futures, so I don't believe this test could ever pass with credentials. Perhaps just have separate checks?

Reply:
😳 Ohh there is a big logical error. I think I should test it for the enabled Pairs only.